### PR TITLE
Matrix Multiply (mmul)

### PR DESCRIPTION
**Pull request by gordoncaleb**
This PR:

- Fixes previous elementwise tensor auto diff to support non-elementwise operations.
- Add matrix multiply vertex and appropriate auto diff for it.

### CODE REVIEW

File: keanu-project/src/test/java/io/improbable/keanu/tensor/dbl/Nd4jDoubleTensorTest.java [hunk 1]
```diff
@@ -134,4 +136,15 @@ public void canApplySqrt() {
         assertEquals(Math.sqrt(2.0), result.scalar(), 0.0);
     }
 
+    @Test
+    public void play() {
```
Comment (GeorgeNash) [inline keanu-project/src/test/java/io/improbable/keanu/tensor/dbl/Nd4jDoubleTensorTest.java]: Looks like you forgot to delete this one

File: keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/binary/MatrixMultiplicationVertexTest.java [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        }, 3, 2));
+
+        DoubleVertex N = alpha.matrixMultiply(m);
+        DoubleVertex L = beta.matrixMultiply(alpha);
+        //y = L x N = (beta x alpha) x (alpha x m)
+        DoubleVertex y = L.matrixMultiply(N);
+        DualNumber yDual = y.getDualNumber();
+
+        DoubleTensor dydalpha = yDual.getPartialDerivatives().withRespectTo(alpha);
+        DoubleTensor expectedDydalpha = DoubleTensor.create(new double[]{
```
Comment (GeorgeNash) [inline keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/binary/MatrixMultiplicationVertexTest.java]: awesome!

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/DualNumber.java [hunk 3]
```diff
@@ -57,29 +57,96 @@ public DualNumber subtract(DualNumber that) {
...
     }
 
+    public DualNumber matrixMultiplyBy(DualNumber that) {
+        // dc = A * db + da * B;
+        DoubleTensor newValue = this.value.matrixMultiply(that.value);
+        PartialDerivatives thisInfMultiplied;
+        PartialDerivatives thatInfMultiplied;
+
+        if (this.partialDerivatives.isEmpty()) {
+            thisInfMultiplied = this.partialDerivatives.clone();
```
[the remark this answers is not in the document]
Comment (gordoncaleb) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/DualNumber.java]: Good call.

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/DualNumber.java [hunk 4]
```diff
@@ -57,29 +57,96 @@ public DualNumber subtract(DualNumber that) {
...
         return new DualNumber(newValue, newInf);
     }
 
     public DualNumber divideBy(DualNumber that) {
         // dc = (B * da - A * db) / B^2;
         DoubleTensor newValue = this.value.div(that.value);
-        PartialDerivatives thisInfMultiplied = this.partialDerivatives.multiplyBy(that.value);
-        PartialDerivatives thatInfMultiplied = that.partialDerivatives.multiplyBy(this.value);
-        PartialDerivatives newInf = thisInfMultiplied.subtract(thatInfMultiplied).divideBy(that.value.times(that.value));
+        PartialDerivatives thisInfMultiplied;
```
[the remark this answers is not in the document]
Comment (gordoncaleb) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/DualNumber.java]: They're actually multiplied. The division magic happens later.

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/PartialDerivatives.java [hunk 5]
```diff
@@ -85,13 +110,103 @@ public PartialDerivatives multiplyBy(DoubleTensor multiplier) {
...
+            int[] partialShape = partial.getValue().getShape();
+            int[] resultShape = Arrays.copyOf(partialShape, partialShape.length);
+
+            DoubleTensor v;
+            if (partialIsLeft) {
+                resultShape[0] = partialShape[0];
+                resultShape[1] = multiplier.getShape()[1];
+                v = partial.getValue()
+                    .tensorMultiply(reshapedMultiplier, new int[]{1}, new int[]{0})
+                    .reshape(-1, resultShape[1])
```
[the remark this answers is not in the document]
Comment (gordoncaleb) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/PartialDerivatives.java]: -1 actually means whatever makes sense given the other dimension is defined. The fact that there's only 2 dimensions specified strips off the extra ones. It's strange and magically but it seems to be the tensor reshape convention (https://stackoverflow.com/questions/18691084/what-does-1-mean-in-numpy-reshape)

File: keanu-project/src/test/java/io/improbable/keanu/tensor/dbl/Nd4jDoubleTensorTest.java (hunk 1 above)
Comment (gordoncaleb) [reply keanu-project/src/test/java/io/improbable/keanu/tensor/dbl/Nd4jDoubleTensorTest.java]: Good catch!

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/PartialDerivatives.java [hunk 6]
```diff
@@ -46,6 +59,18 @@ public void putWithRespectTo(long id, DoubleTensor value) {
         derivativeWithRespectTo.put(id, value);
     }
 
+    public PartialDerivatives sum(int... overDimensions) {
+        Map<Long, DoubleTensor> summed = cloneInfinitesimals(derivativeWithRespectTo);
+
+        for (Map.Entry<Long, DoubleTensor> entry : derivativeWithRespectTo.entrySet()) {
+            long k = entry.getKey();
+            DoubleTensor v = entry.getValue();
+            summed.put(k, reshapeByPrependPad(v.sum(overDimensions), v.getRank()));
```
[the remark this answers is not in the document]
Comment (gordoncaleb) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/PartialDerivatives.java]: I went with:
```
            DoubleTensor reshapedV = v.sum(overDimensions);
            summed.put(k, increaseRankByPrependingOnesToShape(reshapedV, v.getRank()));
```
Watcha think?

File: keanu-project/src/main/java/io/improbable/keanu/tensor/dbl/Nd4jDoubleTensor.java [hunk 7]
```diff
@@ -858,12 +893,12 @@ public void set(long index, Double value) {
 
     @Override
     public double[] asFlatDoubleArray() {
-        return tensor.data().asDouble();
+        return tensor.dup().data().asDouble();
```
[the remark this answers is not in the document]
Comment (GeorgeNash) [reply keanu-project/src/main/java/io/improbable/keanu/tensor/dbl/Nd4jDoubleTensor.java]: Nice, good catch and cheers for the example

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/DualNumber.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (GeorgeNash) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/DualNumber.java]: Nice

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/PartialDerivatives.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (GeorgeNash) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/PartialDerivatives.java]: Yeah, smashed it

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/PartialDerivatives.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (GeorgeNash) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/PartialDerivatives.java]: Okay sweet!

**Pull request merged**